DotNet/Tenant/Controllers/FacilityController.cs

```diff
@@ -271,5 +288,152 @@ public async Task<IActionResult> DeleteFacility(string facilityId, CancellationT
             return NoContent();
         }
 
+        /// <summary>
+        /// Generat
+        /// </summary>
+        /// <param name="facilityId"></param>
+        /// <param name="request"></param>
+        /// <returns></returns>
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        [HttpPost("AdHocReport")]
+        public async Task<IActionResult> GenerateAdHocReport(string facilityId, AdHocReportRequest request)
+        {
+            if (string.IsNullOrEmpty(facilityId) || await _facilityConfigurationService.GetFacilityByFacilityId(facilityId, CancellationToken.None) == null)
+            {
+                return BadRequest("Facility does not exist.");
+            }
+
+            if (request.ReportTypes == null || request.ReportTypes.Count == 0)
+            {
+                return BadRequest("ReportTypes must be provided.");
+            }
+
+            if (request.StartDate == null || request.StartDate == DateTime.MinValue)
+            {
+                return BadRequest("StartDate must be provided.");
+            }
+
+            if (request.EndDate == null || request.EndDate == DateTime.MinValue)
+            {
+                return BadRequest("EndDate must be provided.");
+            }
+
+            if (request.EndDate <= request.StartDate)
+            {
+                return BadRequest("EndDate must be after StartDate.");
+            }
+
+            try
+            {
+                foreach (var rt in request.ReportTypes)
+                {
+                    //this will throw an ApplicationException if the Measure Definition does not exist.
+                    await _facilityConfigurationService.MeasureDefinitionExists(rt);
+                }
+
+                var producerConfig = new ProducerConfig();
+
+                using var producer = _adHocKafkaProducerFactory.CreateProducer(producerConfig);
+
+                var headers = new Headers
+                {
+                    { "X-Report-Tracking-Id", System.Text.Encoding.ASCII.GetBytes(Guid.NewGuid().ToString()) }
+                };
+
+                var message = new Message<string, GenerateReportValue>
+                {
+                    Key = facilityId,
+                    Headers = headers,
+                    Value = new GenerateReportValue
+                    {
+                        ReportId = Guid.NewGuid().ToString(),
```

Comment:
We do not need the reportId here for GenerateAddHoc report. The "X-Report-Tracking-Id" is the one to generate a new scheduleReport record.